Add EC2 provisioning support to Dallinger commandline

alecpm opened the pull request:
This code is derived from @polvanrijn's work on the CAP package. Refs #6588

## Description
Implements and documents various command line extensions related to EC2 instance provisioning. A couple of notes for those already using these commands via the `cap` package:

* I've changed the default security group name from `cap` to `dallinger`, but you can override that in your dallinger config using the `ec2_default_security_group` variable.
* I've changed the default SSH key pair id (PEM file) from `cap` to `dallinger`, but you can override that in your dallinger config using the `ec2_default_pem` variable.
* I've updated the commands to not require a `--region` specification, the `aws_region` config will be used by default if a `--region` is not specified. The exception is the `ec2 list instances` command which will still list instances in all regions when `--region` is not specified.

## Motivation and Context
See #6588

## How Has This Been Tested?
Manual commandline testing. My ability to test provisioning and teardown was limited by some local issues I'm having with experiment deploys, but all other methods were tested.

pmcharrison reviewed (commented):
Thanks @alecpm, I looked through and this looks just right. However I'm unable currently to do a line-by-line review (parental leave). I will ask on Slack if anyone is available for it.

pmcharrison commented on dallinger/command_line/ec2.py:
```diff
@@ -0,0 +1,233 @@
+        exp_klass = experiment.load()
+        exp = exp_klass(db.session)
+
+        asset_storage = getattr(exp, "asset_storage", None)
```
This logic about asset storage can be removed; not only is it PsyNet-specific, but it also is out of date.

alecpm replied:
I was wondering about that. I tried to generalize and decouple it from PsyNet as much as possible, but it's great to hear it can be removed.

jacobyn commented on dallinger/command_line/ec2.py:
```diff
@@ -0,0 +1,233 @@
+        if asset_storage is not None:
+            asset_storage = asset_storage.__class__.__name__
+
+        if asset_storage and not asset_storage == "S3Storage":
```
The S3 part can be removed, right?

jacobyn commented on dallinger/command_line/ec2.py:
```diff
@@ -0,0 +1,233 @@
+                exit()
+    except ImportError:
+        if not user_confirms(
+            "You are not inside an experiment folder. This is not recommended. "
```
This seems not needed, I think this only comes because of Pol making psynet specific tests that anyhow should not be there...

jacobyn commented on docs/source/ec2_deployment.rst:
```diff
@@ -0,0 +1,172 @@
+In order to use EC2 provisioning, you must have a domain created in AWS Route53
+managed using the same credentials you use for your experiment (see:
+:doc:`aws_etc_keys`). The examples below assume that you have setup a domain
+``my-experiments.org`` in Route53.
```
In the Psynet documentation we had the following text, perhaps add something like this here:

"If your lab is doing this for the first time, you probably need to acquire a domain name for your experiment server. This is the parent URL that will be used to host your experiments. If your lab already has a domain name, you can skip this step. On the AWS online console, navigate to the Route 53 service. On the Dashboard you can register a domain name. Note that different domain names come with different costs, and that registering a domain name can take from a few minutes to several hours. Before proceeding with the next steps, please wait until the AWS console tells you that the registration is complete."

This is taken from here were step 15 represent this step: https://psynetdev.gitlab.io/PsyNet/deploy/aws_server_setup.html

jacobyn commented:
Feedback on "functional" test (testing from the perspective of a user)
dallinger --help 
returns no description for  "ec2"

Is there any reason for 
"You are not inside an experiment folder. This is not recommended. Are you sure you want to provision an EC2 instance? Enter 'y' for yes, 'n' for no.
 [y/N]: y"

I think Pol had it because he tested a few things using the experiment class. Considering that this is not really needed to we really need the question?

I made several more comments in the code.

pmcharrison commented:
@jacobyn I agree with not needing the experiment folder. 

For the Docker problem, have you made sure that Rosetta is enabled in your Docker desktop?

alecpm commented:
@jacobyn @pmcharrison I've added help text for the new commands and and command groups, removed the unwanted checks, and added the requested DNS related documentation. This is ready for re-review.

pmcharrison commented on docs/source/ec2_deployment.rst:
```diff
@@ -0,0 +1,180 @@
+-----------------------------------
+
+**Note**: If you decide to deploy to an EC2 instance, you need to "hire" the server
+(Provisioning). Once you do that -- the clock is ticking and you will be charged
```
`--` should be a comma

pmcharrison commented on docs/source/ec2_deployment.rst:
```diff
@@ -0,0 +1,180 @@
+---------------------
+
+You can temporarily stop a running instance to prevent it from incurring hourly
+costs (while retaining it's stored data)::
```
"it's" should be "its"

pmcharrison reviewed (approved):
LGTM!
